thorclient/wsclient/client.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package wsclient
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/gorilla/websocket"
+	"github.com/vechain/thor/v2/api/blocks"
+	"github.com/vechain/thor/v2/api/subscriptions"
+	"github.com/vechain/thor/v2/thorclient/common"
+)
+
+type Client struct {
+	host   string
+	scheme string
+}
+
+func NewClient(url string) (*Client, error) {
+	var host string
+	var scheme string
+
+	if strings.Contains(url, "https://") || strings.Contains(url, "wss://") {
+		host = strings.TrimPrefix(strings.TrimPrefix(url, "https://"), "wss://")
+		scheme = "wss"
+	} else if strings.Contains(url, "http://") || strings.Contains(url, "ws://") {
+		host = strings.TrimPrefix(strings.TrimPrefix(url, "http://"), "ws://")
+		scheme = "ws"
+	} else {
+		return nil, fmt.Errorf("invalid url")
+	}
+
+	return &Client{
+		host:   strings.TrimSuffix(host, "/"),
+		scheme: scheme,
+	}, nil
+}
+
+func (c *Client) SubscribeEvents(query string) (<-chan common.EventWrapper[*subscriptions.EventMessage], error) {
+	conn, err := c.connect("/subscriptions/event", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.EventMessage](conn)
+}
+
+func (c *Client) SubscribeBlocks(query string) (<-chan common.EventWrapper[*blocks.JSONCollapsedBlock], error) {
+	conn, err := c.connect("/subscriptions/block", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[blocks.JSONCollapsedBlock](conn)
+}
+
+func (c *Client) SubscribeTransfers(query string) (<-chan common.EventWrapper[*subscriptions.TransferMessage], error) {
+	conn, err := c.connect("/subscriptions/transfer", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.TransferMessage](conn)
+}
+
+func (c *Client) SubscribeTxPool(query string) (<-chan common.EventWrapper[*subscriptions.PendingTxIDMessage], error) {
+	conn, err := c.connect("/subscriptions/txpool", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.PendingTxIDMessage](conn)
+}
+
+func (c *Client) SubscribeBeats2(query string) (<-chan common.EventWrapper[*subscriptions.Beat2Message], error) {
+	conn, err := c.connect("/subscriptions/beat2", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.Beat2Message](conn)
+}
+
+// subscribe creates a channel to handle new subscriptions
+// It takes a websocket connection as an argument and returns a read-only channel for receiving messages of type T and an error if any occurs.
+func subscribe[T any](conn *websocket.Conn) (<-chan common.EventWrapper[*T], error) {
```

Comment:
Looks like there is no close func for subscriptions, is it intended?

Reply:
Added a `Subscriptions` that handles the closing of the websockets, lmk what you think!

Reply:
I think it might be over engineering? `co.Goes` is for managing sync group, the close function is better managed by the caller.  I think the close function can be easily done in-place in `subscribe` which is clearer.

Another point is the producer won't need to consume all the remaining messages, it's consumer's decision to make. And it is a race condition for both consuming one channel. The producer should right away stop producing new events after receiving the stop signal.

Reply:
I understand. I'm happy with going with another route, let's chat about it offline, but here are the reasoning behind those choices:

- `Subscriptions` as an object
Serves to encapsulate the `Unsubscribe` function (and potentially encapsulate an error channel if required)
I think it's neater to return 1 object and error :
`SubscribeBlocks() (*common.Subscription[*blocks.JSONCollapsedBlock], error)`
vs to return 1 object, one function that closes out the object and error:
`SubscribeBlocks() (*common.EventChan[*blocks.JSONCollapsedBlock], func() error, error)`

- `co.Goes` does not have a stop functionality. `co.Choes` adds a channel that allows to stop all goroutines.
When the `Unsubscribe` function is called, the `co.Choes.stopChan` is signalled to stop execution, making the subscription clean exit.
The caller still has to implement how it listens to the `co.Choes.stopChan` and exits its execution.

- I see what you mean. The comments on the drain messages are not on-point. It's actually only reading one (perhaps a few) event so that the `case <-stopChan:` is picked up. 
Once the `stopChan` is closed, the `eventChan` and the `conn` are closed, and no more messages are read.

What happened before "drainage" was that `err := conn.ReadJSON(&data)` will lock the routine, and it would never reach `case <-stopChan:`. 
I've seen this pattern happen quite a lot because `conn.ReadJSON` is locking.